src/components/vm/consoles/vnc.jsx

```diff
@@ -48,6 +55,79 @@ const Enum = {
     KEY_DELETE: 111,
 };
 
+export const VncState = ({ vm, vnc }) => {
+    const Dialogs = useDialogs();
+
+    function add_vnc() {
+        Dialogs.show(<AddVNC idPrefix="add-vnc" vm={vm} />);
+    }
+
+    function edit_vnc() {
+        Dialogs.show(<EditVNCModal idPrefix="edit-vnc" vm={vm} consoleDetail={vnc} />);
+    }
+
+    if (vm.state == "running" && !vnc) {
+        return (
+            <EmptyState>
+                <EmptyStateBody>
+                    {_("Graphical support not enabled.")}
+                </EmptyStateBody>
+                <EmptyStateFooter>
+                    <Button variant="secondary" onClick={add_vnc}>
+                        {_("Add VNC")}
+                    </Button>
+                </EmptyStateFooter>
+            </EmptyState>
+        );
+    }
+
+    let vnc_info;
+    let vnc_action;
+
+    if (!vnc) {
+        vnc_info = _("not supported");
+        vnc_action = (
+            <Button variant="link" isInline onClick={add_vnc}>
+                {_("Add VNC")}
+            </Button>
+        );
+    } else {
+        if (vnc.port == -1)
+            vnc_info = _("VNC, dynamic port");
+        else
+            vnc_info = cockpit.format(_("VNC, port $0"), vnc.port);
+
+        vnc_action = (
+            <Button variant="link" isInline onClick={edit_vnc}>
+                {_("Edit")}
+            </Button>
+        );
+    }
+
+    return (
+        <>
+            <p>
+                {
+                    vm.state == "running"
+                        ? _("Shut down and restart the virtual machine to access the graphical console.")
+                        : _("Please start the virtual machine to access its console.")
```

Comment:
Again, let's make the sentences consistent. Use "its console" or "the graphical console" both here and in the line above.

Reply:
This line will also be shown when the machine has a serial console but no vnc console. It is already present on "main". So I would leave it as it is.

I think this can all be cleanup and made elegant by also having a switcher when the machine is shut off, as Garrett has proposed. Then each console type can talk only about itself and we don't need to have language that tries to cover all cases.